[DLA 3] Add `qml.dla.adjoint_repr` function

Qottmann opened the pull request:
A function to compute the adjoint representation of a DLA

- [x] Basic functionality
- [x] Basic testing
- [x] Docs
- [x] changelog

[sc-51422]
https://github.com/PennyLaneAI/pennylane/pull/5161 (DLA1) https://github.com/PennyLaneAI/pennylane/pull/5169 (DLA2) https://github.com/PennyLaneAI/pennylane/pull/5406 (DLA3)

dwierichs commented on tests/dla/test_adjoint_repr.py:
```diff
@@ -0,0 +1,69 @@
+Ising3 = qml.dla.lie_closure(gens, pauli=True)
+
+
+class TestAdjointRepr:
```
I think we should have a test where the generators are not single Pauli words, but true linear combinations. Maybe that could just be a second example, which is passed to the `test_adjoint_repr_elements` test?

Qottmann replied:
Good point! This unraveled a lot of things that still needed sorting out, in particular with regards to normalization. Added such a case now and all seems well 👍

dwierichs reviewed (changes requested):
Great job @Qottmann, really concise code with excellent doc and examples!
Again only very minor things, with an abbreviation suggestion and a test request :) 
Very cool to have the `PauliSentence.trace` built in as well!

Qottmann commented on pennylane/dla/adjoint_repr.py:
```diff
@@ -0,0 +1,105 @@
+           [-0., -1., -0.,  0.,  0.,  0.],
+           [-0., -0., -1.,  0.,  0.,  0.],
+           [ 0., -0., -0., -0., -0.,  0.]])
+
```
I think this has become obsolete, at least for the moment, after I removed all normalization now 👍

Qottmann commented:
This shouldnt have taken me that long, but I now removed all normalization and we get consistent results.

When a basis is not normalized, then the vector decomposition is

$$ v = \sum_j \frac{v \cdot e_j}{||e_j||^2} e_j $$

So I added the normalizing factor $1/||e_j||^2$ to the decomposition function

Qottmann commented:
We could consider renaming this function to `structure_constants`

dwierichs commented on tests/dla/test_adjoint_repr.py:
```diff
@@ -0,0 +1,86 @@
+        assert adjoint.shape == (d, d, d)
+        assert adjoint.dtype == float
+
+    @pytest.mark.parametrize("dla", [Ising3, XXZ3])
```
:tada:

dwierichs reviewed (changes requested):
Nice @Qottmann :100:
I think we may discuss the minus sign issue (or my confusion, that is) offline :)

Qottmann commented on pennylane/dla/adjoint_repr.py:
```diff
@@ -0,0 +1,118 @@
+    r"""
+    Compute the adjoint representation of a dynamical Lie algebra (DLA)
+
+    Given a DLA :math:`\{G_1, G_2, .. G_\mathfrak{d} \}` of dimension :math:`\mathfrak{d}`,
```
no particular reason, just to give the dimension of the dla a fancy letter 🙃

dwierichs replied:
Haha nice :) :sunglasses:

Qottmann commented on pennylane/dla/adjoint_repr.py:
```diff
@@ -0,0 +1,118 @@
+    rep = np.zeros((len(dla), len(dla), len(dla)), dtype=float)
+    for i, op in enumerate(dla):
+        for (j, k), res in commutators.items():
+            value = (-1j * (op @ res).trace()).real
```
This minus sign here is coming from $\text{tr}\left(iG_a \cdot [iG_b, iG_c]\right) = -i \text{tr}\left(G_a \cdot [G_b, G_c]\right)$

I think both formulas and examples are correct the way they are right now 😬

Qottmann commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -597,6 +597,21 @@ def __missing__(self, key):
+        0.5
+
+        """
+        return self.get(PauliWord({}), 0.0)
```
nice!

dwierichs commented:
>We could consider renaming this function to structure_constants as that is more apt imo

I wouldn't mind this :)

Qottmann commented on pennylane/dla/adjoint_repr.py:
```diff
@@ -0,0 +1,117 @@
+    rep = np.zeros((len(dla), len(dla), len(dla)), dtype=float)
+    for i, op in enumerate(dla):
+        for (j, k), res in commutators.items():
+            value = (1j * (op @ res).trace()).real
```
see here @vincentmr

Qottmann commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -597,6 +600,21 @@ def __missing__(self, key):
+        0.5
+
+        """
+        return self.get(pw_id, 0.0)
```
All of X, Y, Z are traceless, as well as any combination thereof. The only non-traceless (one may say: traceful lol) operator is the identity. It is true that this is somewhat trivial but this is mathematically handy since the inner product for operators is defined by the trace of their matrix product.

I am using that here because the structure constants are the decomposition coefficients of the commutators, and that decomposition is computed with respect to the appropriate inner product (the (normalized) trace inner product)

See e.g. https://github.com/PennyLaneAI/pennylane/pull/5406/files#r1551936639

Qottmann replied:
So standard vector decomposition $v = \sum c_i e_i$ with $c_i = \langle v, e_i \rangle$, here, in our case, becomes $v \rightarrow [h_\alpha, h_\beta]$ and $e_i \rightarrow h_i$

vincentmr replied:
Got it, thanks for this nice explanation.

vincentmr commented on pennylane/dla/adjoint_repr.py:
```diff
@@ -0,0 +1,112 @@
+    -2.
+
+    We can also look at the overall adjoint action of the first element of the DLA on other elements, :math:`G_0`, ``dla[0] = X(0) @ X(1)``.
+    >>> adjoint_rep[0]
```
Do we need a new line here?

vincentmr commented:
It seems the build is failing because NVIDIA libs are [not found](https://github.com/PennyLaneAI/pennylane/actions/runs/8557913677/job/23451305756?pr=5406#step:5:1051). Should we maybe install them by hand?

Qottmann commented:
> It seems the build is failing because NVIDIA libs are [not found](https://github.com/PennyLaneAI/pennylane/actions/runs/8557913677/job/23451305756?pr=5406#step:5:1051). Should we maybe install them by hand?

That may be an unrelated issue, the sphinx failures came from messed up latex 🙃 now all tests pass 👍

dwierichs commented on pennylane/dla/structure_constants.py:
```diff
@@ -0,0 +1,114 @@
+
+    .. math:: [i G_\alpha, i G_\beta] = \sum_{\gamma = 0}^{d-1} f^\gamma_{\alpha, \beta} iG_\gamma.
+
+    Let us confirm those with some examples. Take :math:`[iG_1, iG_3] = [iZ_0, -iY_0 X_1] = -i 2 X_0 X_1 = -i 2 G_0`, so
```
We're showing mostly one here, really.
```suggestion
    Let us confirm those with an example. Take :math:`[iG_1, iG_3] = [iZ_0, -iY_0 X_1] = -i 2 X_0 X_1 = -i 2 G_0`, so
```

Qottmann commented:
@trbromley I renamed this function `adjoint_repr` -> `structure_constants` as that more aptly describes what the function is doing. I still keep mentioning the adjoint representation for better search results. Are you happy with this change?

trbromley commented on doc/releases/changelog-dev.md:
````diff
@@ -118,6 +118,16 @@
    -1.0 * Y(1) @ Y(0)]
   ```
 
+* We can compute the adjoint representation of a dynamical Lie algebra.
+  [(5406)](https://github.com/PennyLaneAI/pennylane/pull/5406)
+
+  For example, we can compute the adjoint representation of the transverse field Ising model DLA.
+
+  >>> dla = [1.0 * X(0) @ X(1), 1.0 * Z(0), 1.0 * Z(1), -1.0 * Y(0) @ X(1), -1.0 * X(0) @ Y(1), -1.0 * Y(0) @ Y(1)]
+  >>> adjoint_rep = qml.dla.adjoint_repr(dla)
+  >>> adjoint_rep.shape
````
Nice!

trbromley commented on doc/releases/changelog-dev.md:
```diff
@@ -118,6 +118,16 @@
+  For example, we can compute the adjoint representation of the transverse field Ising model DLA.
+
+  >>> dla = [1.0 * X(0) @ X(1), 1.0 * Z(0), 1.0 * Z(1), -1.0 * Y(0) @ X(1), -1.0 * X(0) @ Y(1), -1.0 * Y(0) @ Y(1)]
+  >>> adjoint_rep = qml.dla.adjoint_repr(dla)
```
```suggestion
  >>> adjoint_rep = qml.structure_constants(dla)
```

trbromley commented on pennylane/pauli/dla/structure_constants.py:
```diff
@@ -0,0 +1,125 @@
+    factor :math:`\text{tr}\left( iG_\gamma iG_\gamma \right)` in the projection.
+
+    Args:
+        g (List[Union[Operator, PauliWord, PauliSentence]]): The (dynamical) Lie algebra for which we want to compute
+            its adjoint representation. DLAs can be generated by a set of generators via :func:`~lie_closure`.
```
Out of curiosity, what would happen if you don't put a DLA in here? Would it work but just get unreliable results?

Qottmann replied:
If you put in a list of operators that does not form a Lie algebra, i.e. is not closed under commutation, then the structure constants just become non-sensical. In particular, the assumption $[G_\alpha, G_\beta] = \sum_\gamma f^\gamma_{\alpha \beta} G_\gamma$, that any commutator between elements can be expressed as a linear combination of DLA elements, is wrong. And so the projection onto dla elements (which is what the structure constants are, really) is just incomplete.

trbromley commented on pennylane/pauli/dla/structure_constants.py:
```diff
@@ -0,0 +1,125 @@
+    Let us generate the DLA of the transverse field Ising model using :func:`~lie_closure`.
+
+    >>> n = 2
+    >>> gens = [PauliSentence({PauliWord({i:"X", i+1:"X"}): 1.}) for i in range(n-1)]
+    >>> gens += [PauliSentence({PauliWord({i:"Z"}): 1.}) for i in range(n)]
+    >>> dla = qml.pauli.lie_closure(gens)
+    >>> print(dla)
+    [1.0 * X(0) @ X(1), 1.0 * Z(0), 1.0 * Z(1), -1.0 * Y(0) @ X(1), -1.0 * X(0) @ Y(1), -1.0 * Y(0) @ Y(1)]
```
Very minor, but I wonder if this would look nicer if not using `PauliWord` and `PauliSentence` but instead standard `X` and `Z` ops.

Qottmann replied:
Yes good point! This wasnt possible earlier but is now 👍

trbromley commented on pennylane/pauli/dla/structure_constants.py:
```diff
@@ -0,0 +1,125 @@
+    >>> adjoint_rep[0, 1, 3]
+    -2.
+
+    We can also look at the overall adjoint action of the first element of the DLA on other elements, :math:`G_0`, ``dla[0] = X(0) @ X(1)``.
```
```suggestion
    We can also look at the overall adjoint action of the first element :math:`G_0 = X_{0} \otimes X_{1}` of the DLA on other elements.
```

trbromley commented on pennylane/pauli/dla/structure_constants.py:
```diff
@@ -0,0 +1,125 @@
+    >>> adjoint_rep[0, 1, 3]
+    -2.
+
+    We can also look at the overall adjoint action of the first element of the DLA on other elements, :math:`G_0`, ``dla[0] = X(0) @ X(1)``.
+
+    >>> adjoint_rep[0]
+    array([[ 0.,  0.,  0.,  0.,  0.,  0.],
+           [-0.,  0.,  0., -2.,  0.,  0.],
+           [-0.,  0.,  0.,  0., -2.,  0.],
+           [-0.,  2., -0.,  0.,  0.,  0.],
+           [-0., -0.,  2.,  0.,  0.,  0.],
+           [ 0., -0., -0., -0., -0.,  0.]])
+
+    Note that we neither enforce nor assume normalization by default.
```
I think I'm confused how the indices of `adjoint_rep` correspond to $f$. Is it $f_{\alpha, \beta}^{\gamma} = $`adjoint_rep[alpha, beta, gamma]`?

Qottmann replied:
I was hoping to make that clear in the line above with $f^0_{1, 3}$ referring to `adjoint_rep[gamma, alpha, beta]` but perhaps I should make it more explicit!

Qottmann replied:
✅

trbromley reviewed (commented):
Thanks @Qottmann! I've left a few comments for now. Will give a final look through once the first and second PRs are merged in.
